### PR TITLE
[RDY] Add plugin for editing ShaDa files

### PR DESCRIPTION
**Pull request by ZyX-I**
Wondering whether test/functional/plugin is the right place for testing plugin files.

Note: this is _not_ blocked by #2506 since I do not plan any format changes.

TODO:
- [x] Runtime files.
- [x] autoload/msgpack.vim tests.
- [x] autoload/shada.vim tests.
- [x] ftplugin/shada.vim + plugin/shada.vim tests.
- [x] syntax/shada.vim tests.
- [x] Documentation for autoload/msgpack.vim.
- [x] Documentation for */shada.vim.

### CODE REVIEW

**Comment by ZyX-I** [comment]
Actually it is better to make this blocked by #2506 because I need some helper functions from there for plugin/shada.vim tests (but not for other files).

File: runtime/doc/pi_msgpack.txt [hunk 1]
```diff
@@ -0,0 +1,139 @@
+*pi_msgpack.txt*	For NeoVim version 7.4.
```
Comment (fmoralesc) [inline runtime/doc/pi_msgpack.txt]: I don't think the version is appropriate.

File: runtime/doc/pi_msgpack.txt [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	equal to `%Y-%m-%dT%H:%M:%S`.
+
+msgpack#int_dict_to_str({msgpack-special-int})	*msgpack#int_dict_to_str()*
+	Function which converts |msgpack-special-dict| integer value to 
+	a hexadecimal value like 0x1234567890ABCDEF (always returns exactly 16 
+	hexadecimal digits).
+
+msgpack#special_type({msgpack-value})		*msgpack#special_type()*
+	Returns zero if {msgpack-value} is not |msgpack-special-dict|.  If it 
+	is returns name of the key in |v:msgpack_types| which represents
```
Comment (fmoralesc) [inline runtime/doc/pi_msgpack.txt]: is returns -> is, it returns

File: runtime/doc/pi_msgpack.txt [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	   msgpack#deepcopy function.............: |msgpack#deepcopy()|
+	   msgpack#string function...............: |msgpack#string()|
+	   msgpack#eval function.................: |msgpack#eval()|
+	   msgpack#equal function................: |msgpack#equal()|
+
+
+==============================================================================
+2. Msgpack.vim introduction			*msgpack.vim-intro*
+
+This plugin contains utility functions used in conjunction with
```
Comment (fmoralesc) [inline runtime/doc/pi_msgpack.txt]: used -> to be used

**Comment by ZyX-I** [comment]
So “Update branch” button adds merge commit (merged master into my branch). It is interesting whether merging it afterwards will produce another merge commit or not.

**Comment by dnmario** [comment]
Dear Nikolai Aleksandrovich Pavlov,

The email sent to danielm@tinyhub.tk with subject
"Re: [neovim] [WIP] Add plugin for editing ShaDa files (#3270)"
will be delivered as soon as possible.

Your email address is automatically added to the
recipients 'whitelist'. This means for further
emails to danielm@tinyhub.tk, you will not receive a delivery
notification again.

TK MAILIAS
*************************************************
Looking for a way to Protect Your Identity and 
for a great no-more-spam solution for your email
address reply@reply.github.com? Sign up for TK Mailias.

It's free! Please visit http://www.mail.tk/

**Comment by ZyX-I** [comment]
(I removed the merge commit by force-push, so you will not see it here.)

**Comment by ZyX-I** [comment]
This PR is finished, but SourceCmd event cannot be tested without #2506 and there is also some common code.

**Comment by justinmk** [comment]
I haven't had any problems with #2506 since you fixed the last memory access issue. I think we should aim to merge it by Monday.

**Comment by tarruda** [comment]
@ZyX-I the #3334 fix is merged, you can rebase to make the tests pass.

**Comment by ZyX-I** [comment]
I have updated the PR and it is now ready for merging.

**Comment by ZyX-I** [comment]
Wondering why it fails with

```
make: *** No rule to make target `unittest'.  Stop.
make: Leaving directory `/home/travis/build/neovim/neovim/build'
```

And non-failing builds did not run unit or functional tests at all, only old tests.

**Comment by ZyX-I** [comment]
And #3581 does run unit and functional tests. Will try restarting the whole build.

**Comment by fmoralesc** [comment]
@ZyX-I I found that locally too, wonder what is going on.

**Comment by ZyX-I** [comment]
@fmoralesc Unable to reproduce this locally. Also I see that https://travis-ci.org/neovim/neovim/jobs/88665063 is no longer affected.

**Comment by fmoralesc** [comment]
Maybe a temporary thing. This is the sort of thing I've seen happen when
travis messes up somehow.

**Comment by fmoralesc** [comment]
Bump. Anything else to be done here?

**Comment by ZyX-I** [comment]
Nothing. I have already reported this as ready in https://github.com/neovim/neovim/pull/3270#issuecomment-148777470.

**Comment by fmoralesc** [comment]
OK, then. It looks good. I'll wait for @justinmk or @tarruda to comment, but I think we might just merge this later today (and then, #3581)

**Comment by mhinz** [comment]
:+1:

**Pull request merged**